### PR TITLE
New temporal date-time builtin functions (DMN 1.3)

### LINKED ISSUE #180: Add new temporal built-in functions
**Is your feature request related to a problem? Please describe.**
DMN 1.3 introduced new temporal built-in functions that are useful when working with a date or a date-time value.

**Describe the solution you'd like**
Add the following temporal built-in functions that were introduced with DMN 1.3:
* `day of year( date )`
* `day of week( date )`
* `month of year( date )`
* `week of year( date )`

image

image

**Related issues**

* Camunda BPM: 
* Zeebe broker:

### PR DESCRIPTION
**Pull request by saig0**
## Description

* add the following built-in functions: 
  * `week of year()`
  * `month of year()`
  * `day of week()`
  * `day of year()`

## Related issues

closes #180

### CODE REVIEW

File: docs/develop/feel-built-in-functions.md [hunk 1]
```diff
@@ -826,3 +826,55 @@ Returns the current date.
...
+* result: number
+
+```js
+day of year(date("2019-09-17"))
+// 260
+```
+
+### week of year()
+
+Returns the Gregorian number of the week within the year, accordingly to ISO 8601.
```
Comment (korthout) [inline docs/develop/feel-built-in-functions.md:858]:
```suggestion
Returns the Gregorian number of the week within the year, according to ISO 8601.
```

File: src/test/scala/org/camunda/feel/impl/builtin/BuiltinTemporalFunctionsTest.scala [hunk 2]
```diff
@@ -46,4 +51,38 @@ class BuiltinTemporalFunctionsTest
...
+
+  "The week of year() function" should "return the number of week within the year" in {
+
+    eval(s"week of year($date)") should be(ValNumber(38))
+    eval(s"week of year($localDateTime)") should be(ValNumber(38))
+    eval(s"week of year($dateTime)") should be(ValNumber(38))
+
+    eval(""" week of year(date(2003,12,29)) """) should be(ValNumber(1))
+    eval(""" week of year(date(2004,1,4)) """) should be(ValNumber(1))
+    eval(""" week of year(date(2005,1,1)) """) should be(ValNumber(53))
```
Comment (korthout) [inline src/test/scala/org/camunda/feel/impl/builtin/BuiltinTemporalFunctionsTest.scala:83]: Wait, is this real? 53 weeks in 2005? 😆

File: src/test/scala/org/camunda/feel/impl/builtin/BuiltinTemporalFunctionsTest.scala [hunk 3]
```diff
@@ -46,4 +51,38 @@ class BuiltinTemporalFunctionsTest
       eval(""" today() """) should be(ValDate(now.toLocalDate))
   }
 
+  "The day of year() function" should "return the day within the year" in {
+
+    eval(s"day of year($date)") should be(ValNumber(260))
+    eval(s"day of year($localDateTime) ") should be(ValNumber(260))
+    eval(s"day of year($dateTime) ") should be(ValNumber(260))
```
Comment (korthout) [inline src/test/scala/org/camunda/feel/impl/builtin/BuiltinTemporalFunctionsTest.scala:58]: Considering the edge cases for `week of year`. Perhaps it makes sense to add some edge cases here too. Years with more or less than 365 days (if that exists).

Comment (korthout) [reply src/test/scala/org/camunda/feel/impl/builtin/BuiltinTemporalFunctionsTest.scala:58]: Also, can we assume that a `date/datetime` object is always a real date. For example does `date(2020,9,40)` become `null` because it could not be parsed, or does it become a real date object?

File: src/main/scala/org/camunda/feel/impl/builtin/TemporalBuiltinFunctions.scala [hunk 4]
```diff
@@ -29,4 +37,37 @@ class TemporalBuiltinFunctions(clock: FeelEngineClock) {
...
+    )
+
+  private def getDayOfYear(date: Date): ValNumber = {
+    val dayOfYear = date.getDayOfYear
+    ValNumber(dayOfYear)
+  }
+
+  private def getDayOfWeek(date: Date): ValString = {
+    val dayOfWeek = date.getDayOfWeek
+    val displayName = dayOfWeek.getDisplayName(TextStyle.FULL, Locale.ENGLISH)
```
Comment (korthout) [inline src/main/scala/org/camunda/feel/impl/builtin/TemporalBuiltinFunctions.scala:57]: Did you consider overloading the function with an additional Locale parameter (or a String that can be parsed to a Locale)?

**Review by korthout — APPROVED** [review]
Awesome change, very useful build-in functions. Already LGTM, but I do have some questions.

File: src/test/scala/org/camunda/feel/impl/builtin/BuiltinTemporalFunctionsTest.scala (hunk 3 above)
Comment (saig0) [reply src/test/scala/org/camunda/feel/impl/builtin/BuiltinTemporalFunctionsTest.scala:58]: Yes. It would fail to create this date and return `null` instead.

File: src/main/scala/org/camunda/feel/impl/builtin/TemporalBuiltinFunctions.scala (hunk 4 above)
Comment (saig0) [reply src/main/scala/org/camunda/feel/impl/builtin/TemporalBuiltinFunctions.scala:57]: Good idea :+1: 
But let's wait until someone requests it :smiling_imp:
